avr/variants/sanguino/pins_arduino.h

```diff
@@ -100,9 +102,14 @@ static const uint8_t A7 = PIN_A7;
 #define NUM_ANALOG_INPUTS           8
 
 #define analogInputToDigitalPin(p)  ((p < 8) ? 31 - (p): -1)
-#define analogPinToChannel(p)       ((p < 8) ? (p) : 31 - (p))
+#define analogPinToChannel(p)       ((p) < 8 ? (p) : (p) >= 24 ? 31 - (p) : -1)
 
-#define digitalPinHasPWM(p)         ((p) == 3 || (p) == 4 || (p) == 12 || (p) == 13 || (p) == 14 || (p) == 15 )
+#if defined(__AVR_ATmega644__) || defined(__AVR_ATmega644P__)
+#define digitalPinHasPWM(p)         ((p) == 3 || (p) == 4 || (p) == 12 || (p) == 13 || (p) == 14 || (p) == 15)
+
+#elif defined(__AVR_ATmega1284__) || defined(__AVR_ATmega1284P__)
+ #define digitalPinHasPWM(p)        ((p) == 3 || (p) == 4 || (p) == 6 || (p) == 7 || (p) == 12 || (p) == 13 || (p) == 14 || (p) == 15)
+#endif
```

Comment:
Also this could check if TCCR3A is defined like so:

```C++
#ifdef TCCR3A
#define digitalPinHasPWM(p)    ((p) == 3 || (p) == 4 || (p) == 6 || (p) == 7 || (p) == 12 || (p) == 13 || (p) == 14 || (p) == 15)
#else
#define digitalPinHasPWM(p)    ((p) == 3 || (p) == 4 || (p) == 12 || (p) == 13 || (p) == 14 || (p) == 15)
#endif
```